keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -9,105 +9,201 @@
 
 ## Motivation
 
-For operators who want to regularly rotate the signing and verifying keys for projected volume tokens, the Kubernetes API server must be restarted in order to use a new key. To facilitate easy key rotation, this KEP includes an proposal for a grpc API to support out of process signing and listing of signing keys.
+Enable integration with alternate key management solutions for service account keys. While at it, make the association between kube-apiserver and service account keys more agile.
+
+Benefits are 2 fold:
+1. Ease of rotation.
+  At present, kube-apiserver loads service account keys at process start time and the key remains the same for the life-time of the process. Thus, rotating keys require kube-apiserver restart. External key management eliminates the need to restart.
+2. Enhanced security.
+  kube-apiserver loads service account keys from a file on the disk. Which means that the service account keys are always present on the disk and can be easily accessed by the cluster administrators. With an external signer, we do not need to store the service account keys on the disk. 
 
 ### Goals
 
-- Support for out-of-process JWT signing
-- Support for listing public verifying keys
+- Enable integration with HSMs and cloud KMSes
+- Support out-of-process JWT signing
+- Support listing public verifying keys
+- Prevent external signers from controlling the claims embedded in JWT
 - Preserve existing behavior and performance for keys not read over a socket
+<<[UNRESOLVED]>>
+- Move legacy token signing from kube-controller-manager to kube-apiserver and thus external signing.
+<<[/UNRESOLVED]>>
 
 ### Non-Goals
 
-- Reading TLS serving certificates and key from a socket or reloading of the API server with new cert and key
+- Reading TLS serving certificates and key from a socket or reloading of kube-apiserver with new cert and key
 - Reading any other certificates from a file
 
 ## Proposal
 
 ### Preserve existing behavior
 
-The API server flags `--service-account-key-file` and `--service-account-signing-key-file` will continue be used for reading from files.
+The kube-apiserver flags `--service-account-key-file` and `--service-account-signing-key-file` will continue to be used for reading from files.
 
-### Updates to API server token generation
+### New API
 
-As of Kubernetes v1.13.2, the API server uses the functions `JWTTokenGenerator` and `JWTTokenAuthenticator`. New types that implement the `TokenGenerator` interface and support token validation will be added to `k8s.io/kubernetes/pkg/serviceaccount/`.
+I'm proposing creating a new versioned grpc API (ExternalJWTSigner) under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`. It will have methods to sign JWTs and fetch supported keys (proto in [Design Details](#design-details) section). 
```

Comment:
```suggestion
A new versioned grpc API (ExternalJWTSigner) will be created under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`. It will have methods to sign JWTs and fetch supported keys (proto in [Design Details](#design-details) section). 
```